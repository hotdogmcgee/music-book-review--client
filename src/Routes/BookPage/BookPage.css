#BookPage {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 0 20px;
}

.BookPage__book-info {
  width: 100%;
}

.BookPage__book-info h3 {
  text-align: center;
}
.BookPage__review-list {
  list-style: none;
  padding-left: 0;
}


.BookPage__ratings-container {
  font-size: 16px;
  display: flex;
  flex-direction: column;
  margin-top: 30px;
  align-items: center;
}

.BookPage__review {
  display: flex;
  flex-direction: column;
  border-bottom: 1px dotted black;
}

.BookPage__review:last-child {
  border-bottom: none
}

.BookPage__review-user {
  font-size: 12px;
}

.BookPage__authors-container {
  font-size: 16px;
}

.BookPage__reviews-container {
  width: calc(80vw);
}

.BookPage__authors-container-wide {
  display: none;
}

@media screen and (min-width: 850px) {
  #BookPage {
    padding: 0 100px;
  }

  .BookPage__description {
    line-height: 1.5em;
  }

  .BookPage__book-info h3 {
    font-size: 28px
  }

  .BookPage__authors-container-wide {
    display: block;
    font-size: 12px;
    margin-top: 3px
  }

  .BookPage__authors-container {
    display: none;
  }
}

@media screen and (min-width: 1500px) {
  #BookPage {
    max-width: 80%
  }
}

