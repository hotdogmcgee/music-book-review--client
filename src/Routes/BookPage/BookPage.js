import React from "react";
import { Section, Hyph, NiceDate } from "../../Components/Utils/Utils";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { BookStarRating } from "../../Components/BookStarRating/BookStarRating";
import BookContext from "../../Contexts/BookContext";
import ReviewForm from "../../Components/ReviewForm/ReviewForm";
import BooksApiService from "../../services/books-api-service";
import ReviewSuccessModal from "../../Components/ReviewSuccessModal/ReviewSuccessModal";
import ErrorModal from "../../Components/ErrorModal/ErrorModal";
import "./BookPage.css";

export default class BookPage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      hasSubmission: false,
      showModal: false,
      showErrorModal: false
    };
  }

  static defaultProps = {
    match: { params: {} }
  };

  static contextType = BookContext;

  componentDidMount() {
    const { bookId } = this.props.match.params;
    BooksApiService.getBook(bookId)
      .then(this.context.setBook)
      .catch(this.context.setError);
    BooksApiService.getBookReviews(bookId)
      .then(this.context.setReviews)
      .catch(this.context.setError);
  }

  componentWillUnmount() {
    this.context.clearBook();
  }

  showModal = () => {
    this.setState({ showModal: true });
  };

  hideModal = () => {
    this.setState({ showModal: false });
  };

  showErrorModal = () => {
    this.setState({ showErrorModal: true });
  };

  hideErrorModal = () => {
    this.setState({ showErrorModal: false });
  };

  handleReviewFailure = () => {
    return this.showErrorModal();
  };
  handleReviewSuccess = () => {
    this.setState({ showModal: true });
  };

  renderBook() {
    if (!this.context.book) {
      return "";
    }

    const {
      title,
      authors,
      instrument,
      avg_rating,
      description,
      image,
      num_reviews,
      user
    } = this.context.book;

    const { reviews } = this.context;

    const roundedNumber = avg_rating ? avg_rating.toFixed(1) : 0;

    return (
      <Section id="BookPage">
        <p>{image}</p>
        <div className="BookPage__book-info">
          <h3>{title}<span className="BookPage__authors-container-wide"> by <RenderAuthors authors={authors} /></span></h3>

          <div className="BookPage__authors-container">
            by {authors ? <RenderAuthors authors={authors} /> : "no authors!"}
          </div>

          {/* make this a link to view instrument type? */}
          <p className="capitalize">{instrument}</p>
          <p className="BookPage__description">{description}</p>
        </div>
        <div className="BookPage__ratings-container">
          <BookStarRating rating={roundedNumber} />
          <span>Based on {num_reviews} reviews</span>
        </div>
        <div className="BookPage__reviews-container">
          <BookReviews reviews={reviews} user={user} />
        </div>

        <div className="BookPage__review-form-container">
          <ReviewForm
            onReviewSuccess={this.handleReviewSuccess}
            onReviewFailure={this.handleReviewFailure}
          />
        </div>
      </Section>
    );
  }

  render() {
    return (
      <>
        {this.renderBook()}
        <ReviewSuccessModal
          handleClose={this.hideModal}
          show={this.state.showModal}
          bookId
        />
        <ErrorModal
          error={this.context.error}
          handleClose={this.hideErrorModal}
          show={this.state.showErrorModal}
        />
      </>
    );
  }
}

function BookReviews({ reviews = [] }) {
  return (
    <ul className="BookPage__review-list">
      {reviews.map(review => (
        <li key={review.id} className="BookPage__review">
          <p className="BookPage__review-text">
            <FontAwesomeIcon
              size="sm"
              icon="quote-left"
              className="BookPage__review-icon blue"
            />{" "}
            {review.review_text}
          </p>
          <p className="BookPage__review-user">
            <BookStarRating rating={review.rating} />
            <Hyph />
            {review.user.full_name}
            <Hyph />
            <NiceDate date={review.date_created} />
          </p>
        </li>
      ))}
    </ul>
  );
}

function RenderAuthors({ authors }) {
  return authors.map((author, key) => {
    const commaSpace = key < authors.length - 1 ? ", " : "";

    return (
      <span className="author-container" key={key}>
        {`${author.first_name} ${author.last_name}${commaSpace} `}
      </span>
    );
  });
}
