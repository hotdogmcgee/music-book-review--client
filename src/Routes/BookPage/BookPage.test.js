import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import BookPage from './BookPage'

describe.only(`BookPage component`, () => {
  const props = {
    image: 'an image',
            title: 'where the red fern grows',
            author: 'i forgot',
            description: 'two dogs in a very sad tale',
            rating: 4,
            numReviews: 10,
            id: 1,
            reviews: [
                {
                    user: {
                        full_name: 'dave bones'
                    },
                    rating: 3,
                    id: 1,
                    text: 'it was pretty good'
                }
            ]
  }

  it('renders a BookPage by default', () => {
    const wrapper = shallow(<BookPage />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('renders the Book given props', () => {
    const wrapper = shallow(<BookPage {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})