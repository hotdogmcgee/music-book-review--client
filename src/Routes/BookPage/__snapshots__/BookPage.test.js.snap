// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookPage component renders a BookPage by default 1`] = `
<Fragment>
  <Link
    to="/"
  >
    Music Book Review
  </Link>
  <Section
    id="book"
  >
    <p />
    <div
      className="book-info"
    >
      <h3 />
      <p />
      <p />
    </div>
    <SaveBook />
    <div
      className="ratings-container"
    >
      <span />
      <span>
        Based on 
         reviews
      </span>
    </div>
    <BookReviews />
    <ReviewForm />
  </Section>
</Fragment>
`;

exports[`BookPage component renders the Book given props 1`] = `
<Fragment>
  <Link
    to="/"
  >
    Music Book Review
  </Link>
  <Section
    id="book"
  >
    <p>
      an image
    </p>
    <div
      className="book-info"
    >
      <h3>
        where the red fern grows
      </h3>
      <p>
        i forgot
      </p>
      <p>
        two dogs in a very sad tale
      </p>
    </div>
    <SaveBook />
    <div
      className="ratings-container"
    >
      <span>
        4
      </span>
      <span>
        Based on 
        10
         reviews
      </span>
    </div>
    <BookReviews
      reviews={
        Array [
          Object {
            "id": 1,
            "rating": 3,
            "text": "it was pretty good",
            "user": Object {
              "full_name": "dave bones",
            },
          },
        ]
      }
    />
    <ReviewForm />
  </Section>
</Fragment>
`;
